_posts/2021-07-13-JEV.md

```diff
@@ -0,0 +1,822 @@
+a `digital logic gate` on steroids, that can do A LOT of ~~dumb~~ simple things really REALLY fast. 
+
+##### Vivisection
+
+To give you an idea of what "simple" means, if you want to add 2 numbers you have to:
+
+- 1st simple operation (`mov reg1, num1`): load first number into register (small memory inside CPU)
+- 2nd simple operation (`add reg1, num2`): add second number to register
+- 3rd simple operation (`mov num1, reg1`): write register back to first number
+
+That is 3 CPU operations for something that seems rather innocent. The good news is, is that your CPU (depending on
+speed, and the type of ops) can do millions of these operations per second.
+
+Let's say our CPU can do a million operations per second. Let us do a simplified dissection of the first example from
+above:
+
+- 3 instructions: save current state
+- 3 instructions: setting up the loop
+    - 3 instructions: checking/incrementing the loop
+    - 3 instructions: adding 1 to a value (`mov, add, mov`)
+- 3 instructions: restore saved state
+
+So about 6 instructions per loop iteration. Multiply that by the size of our loop `6 x 6220800 = 37324800`, which when
+divided by a million operations per second results in `37324800 / 1000000 = 37` seconds to increase the brightness of
+a `1920 x 1080` picture (on a fundamental level, computers have a mechanical nature to them).
+
+If you think this sounds ridiculous dear reader, know that it is the truth, albeit half the truth. Your computer truly
+does operate in this way, but that does not mean that this is the **only** way it can operate. If changing every pixel
+on a screen took 37 seconds, then neither video-games nor video playback would be possible on a computer.
+
+#### Soothing light at the end the tunnel
+
+It might surprise/scare you to know, that a lot of the time, software boils down to loads of repetitive operations like
+the ones we described above (disclaimer: this depends of course, but it's our story, so it is true). The point being, our
+example is not an edge case. Meaning that a solution is needed, which in turn results in companies inventing ways to
+address the ~~business opportunity~~ problem.
+
+##### Enter the SIMD
+
+Again, etymology and history are available on the Wikipedia, so forgive me for skipping them. SIMD stands for **Single
+Instruction Multiple Data**. Let's put a pin in that, and recap our problem.
+
+###### The nail to SIMD's Hammer
+
+- In our example, it takes us `37324800` instructions to add 1 to every element of our array.
+- The instructions are very repetitive
+- The instruction sequences are also very repetitive
+- Almost half the instructions are **checking/incrementing the loop**
+- And finally, the array elements do not depend on each other
+
+Okay, so common sense would dictate that our program would run faster in 1 of 2 cases:
+
+- Less instructions
+- Faster instructions - we won't go into this, but some CPU instructions are faster than others. For this example, we
```

Comment:
SIMD instructions have to be fast as well of course, not that their overhead thrashes the rest (I guess it is evident, but perhaps could be made explicit).